src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -656,15 +549,27 @@ struct
         add_gates public_gates ;
         add_gates gates ;
 
+        (* compute the circuit's digest *)
+        let digest = Gates.digest rust_gates in
+        let md5_digest = Md5.digest_bytes digest in
+
         (* drop the gates, we don't need them anymore *)
-        sys.gates <- `Finalized ;
+        sys.gates <- `Finalized (md5_digest, rust_gates) ;
 
         (* return the gates *)
         rust_gates
 
   (** Calls [finalize_and_get_gates] and ignores the result. *)
   let finalize t = ignore (finalize_and_get_gates t : Gates.t)
 
+  (* Returns a hash of the circuit. *)
+  let rec digest (sys : t) =
+    match sys.gates with
+    | `Unfinalized_rev _ ->
+        finalize sys ; digest sys
+    | `Finalized (digest, _) ->
+        digest
```

Comment:
`digest` is sometimes called before we finalize a circuit (at least in the tests), if this happens we finalize first and then recursively call `digest` again